chain/indexer/src/snapshot/download.rs

```diff
@@ -0,0 +1,414 @@
+///
+/// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
+/// let downloader = SnapshotDownloader::new();
+///
+/// // Download from HTTPS
+/// downloader
+///     .download_snapshot(
+///         "https://snapshots.hoprnet.org/logs.tar.gz",
+///         Path::new("/tmp/snapshot.tar.gz"),
+///     )
+///     .await?;
+///
+/// // Copy from local file
+/// downloader
+///     .download_snapshot("file:///backups/snapshot.tar.gz", Path::new("/tmp/snapshot.tar.gz"))
+///     .await?;
+/// # Ok(())
+/// # }
+/// ```
+pub struct SnapshotDownloader {
+    client: Client,
+    config: DownloadConfig,
+}
+
+impl SnapshotDownloader {
+    /// Creates a new snapshot downloader with default configuration
+    pub fn new() -> Self {
+        Self::with_config(DownloadConfig::default())
+    }
+
+    /// Creates a new snapshot downloader with custom configuration
+    pub fn with_config(config: DownloadConfig) -> Self {
+        Self {
+            client: Client::builder()
+                .timeout(config.timeout)
+                .build()
+                .expect("Failed to create HTTP client"),
+            config,
+        }
+    }
+
+    /// Downloads a snapshot from the given URL to the target path.
+    ///
+    /// Supports HTTP/HTTPS URLs with retry logic and file:// URLs for local files.
+    ///
+    /// # Arguments
+    ///
+    /// * `url` - Source URL (http://, https://, or file:// scheme)
+    /// * `target_path` - Destination file path
+    ///
+    /// # Errors
+    ///
+    /// Returns [`SnapshotError`] for network failures, file system errors, or validation failures.
+    pub async fn download_snapshot(&self, url: &str, target_path: &Path) -> SnapshotResult<()> {
+        self.download_snapshot_with_retry(url, target_path, self.config.max_retries)
+            .await
+    }
+
+    /// Downloads a snapshot with configurable retry logic.
+    ///
+    /// Implements exponential backoff between retry attempts for HTTP/HTTPS URLs.
+    /// Local file:// URLs are handled without retry logic. Certain errors
+    /// (like 4xx HTTP status codes or insufficient disk space) will not be retried.
+    ///
+    /// # Arguments
+    ///
+    /// * `url` - The HTTP/HTTPS or file:// URL to download/copy from
+    /// * `target_path` - Local path where the downloaded file will be saved
+    /// * `max_retries` - Maximum number of retry attempts (ignored for file:// URLs)
+    ///
+    /// # Errors
+    ///
+    /// Returns `SnapshotError` for various failure conditions including:
+    /// - Network errors (with retry)
+    /// - HTTP errors (4xx without retry, 5xx with retry)
+    /// - Insufficient disk space (without retry)
+    /// - File size exceeding limits (without retry)
+    pub async fn download_snapshot_with_retry(
+        &self,
+        url: &str,
+        target_path: &Path,
+        max_retries: u32,
+    ) -> SnapshotResult<()> {
+        let mut last_error = None;
+
+        for attempt in 0..=max_retries {
```

Comment:
For this there's a crate called `backon` that we're already using. Check it out.